### PR TITLE
rshim-user-space: add bfb-install

### PR DESCRIPTION
**Pull request by secufoe**
Make bfb-install script available for convenient bootstream installation including log output and progress feedback.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thillux** [comment]
I'd recommend to add an attribute like enableBfbInstall, enable it by default and make the bfb-install dependencies conditional.

File: pkgs/tools/misc/rshim-user-space/default.nix [hunk 1]
```diff
@@ -24,12 +27,15 @@ stdenv.mkDerivation rec {
...
     pkg-config
+    makeBinaryWrapper
   ];
 
   buildInputs = [
     pciutils
     libusb1
     fuse
+    busybox
+    pv
```
Comment (nikstur) [inline pkgs/tools/misc/rshim-user-space/default.nix:38]: This shouldn't be needed if you wrap them explicitly.

**Review by nikstur — COMMENTED** [review]
Looks good. Have you tested bfb-install on a BlueField?

**Comment by thillux** [comment]
> Looks good. Have you tested bfb-install on a BlueField?

I've tested on a host with bluefield attached. The current version works for me.

```
result/bin/bfb-install --bfb ~/markus-installer.bfb --rshim rshim0
Pushing bfb
 484MiB 0:01:10 [6.90MiB/s] [                     <=>                          ]
Collecting BlueField booting status. Press Ctrl+C to stop…
 INFO[BL2]: start
 INFO[BL2]: DDR POST passed
 INFO[BL2]: UEFI loaded
 INFO[BL31]: start
 INFO[BL31]: lifecycle GA Secured
 INFO[BL31]: runtime
 INFO[UEFI]: eMMC init
 INFO[UEFI]: UPVS valid
 INFO[UEFI]: eMMC probed
 INFO[UEFI]: PMI: updates started
 INFO[UEFI]: PMI: boot image update
 INFO[UEFI]: PMI: updates completed, status 0
 INFO[UEFI]: PCIe enum start
 INFO[UEFI]: PCIe enum end
 INFO[UEFI]: exit Boot Service

^C
```

**Comment by thillux** [comment]
I can test again after the small refactoring mentioned above is done.

**Comment by secufoe** [comment]
> I'd recommend to add an attribute like enableBfbInstall, enable it by default and make the bfb-install dependencies conditional.

@nikstur do you agree with having a switch for bfb-install?

**Comment by nikstur** [comment]
> @nikstur do you agree with having a switch for bfb-install?

Sure seems reasonable. I'd prefer `withBfbInstall` though but the exact wording is not super important.

**Comment by secufoe** [comment]
i retested with an attached Bluefield after making bfb-install optional:
```
bfb-install  -r rshim1 -b installer.bfb 
Pushing bfb
1.26GiB 0:02:23 [9.00MiB/s] [                                                                                   <=>                                 ]
Collecting BlueField booting status. Press Ctrl+C to stop…
 INFO[BL2]: start
 INFO[BL2]: DDR POST passed
 INFO[BL2]: UEFI loaded
 INFO[BL31]: start
 INFO[BL31]: lifecycle GA Secured
 INFO[BL31]: runtime
 INFO[UEFI]: eMMC init
 INFO[UEFI]: UPVS valid
 WARN[UEFI]: UPVS full
 WARN[UEFI]: UPVS reclaim
 INFO[UEFI]: eMMC probed
 WARN[UEFI]: Var reclaim
 WARN[UEFI]: Var reclaim done
 INFO[UEFI]: PMI: updates started
 INFO[UEFI]: PMI: boot image update
 INFO[UEFI]: PMI: updates completed, status 0
 INFO[UEFI]: PCIe enum start
 INFO[UEFI]: PCIe enum end
 INFO[UEFI]: exit Boot Service
```

File: pkgs/tools/misc/rshim-user-space/default.nix (hunk 1 above)
Comment (secufoe) [reply pkgs/tools/misc/rshim-user-space/default.nix]: removed it

**Comment by thillux** [comment]
LGTM

**Review by thillux — APPROVED** [review]

**Comment by nikstur** [comment]
Result of `nixpkgs-review pr 261376` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>rshim-user-space</li>
  </ul>
</details>

**Review by nikstur — APPROVED** [review]

**Pull request merged**